include/seqan3/io/record.hpp

```diff
@@ -0,0 +1,312 @@
+ *
+ * // record is tuple-like type, but allows access via field identifiers:
+ * auto & id = get<field::ID>(record);
+ * auto & seq = get<field::SEQ>(record);
+ * ```
+ *
+ */
+template <field ...fs>
+struct fields
+{
+    //!\privatesection
+    //!\brief The template parameters stored in an array for easy access.
+    static constexpr std::array<field, sizeof...(fs)> as_array{fs...};
+
+    //!\brief Retrieve the position of field in the parameter pack.
+    static constexpr size_t index_of(field f)
+    {
+        for (size_t i = 0; i < sizeof...(fs); ++i)
+            if (as_array[i] == f)
+                return i;
+        return std::numeric_limits<size_t>::max();
+    }
+
+    //!\brief Whether a field is contained in the parameter pack.
+    static constexpr bool contains(field f)
+    {
+        return index_of(f) != std::numeric_limits<size_t>::max();
+    }
+
+    static_assert([] () constexpr
+                  {
+                      for (size_t i = 0; i < as_array.size(); ++i)
+                          for (size_t j = i + 1; j < as_array.size(); ++j)
+                              if (as_array[i] == as_array[j])
+                                  return false;
+
+                      return true;
+                  } (), "You may not include a field twice into fields<>.");
+};
+
+} // namespace seqan3
+
+// ----------------------------------------------------------------------------
+// generate_record_tuple_type
+// ----------------------------------------------------------------------------
+
+namespace seqan3::detail
+{
+
+/*!\brief A type metafunction that generates the tuple type for internal use in seqan3::record.
+ * \ingroup io
+ * \tparam field_types          The types of the fields available to the record in a meta::list.
+ * \tparam field_types_as_ids   A seqan3::fields type with seqan3::field IDs corresponding to field_types.
+ * \tparam selected_field_ids   A seqan3::fields type with the subset (and order) of the fields selected.
+ * \tparam tup_t                The tuple type being aggregated.
+ * \tparam field_no             The field we are currently processing.
+ *
+ * This metafunction "iterates" over the selected_field_ids and retrieves the corresponding typenames from
+ * field_types via their identifer in field_types_as_ids. It recursively builds up tup_t which it returns
+ * once the end of selected_field_ids is reached.
+ */
+template <typename field_types,
+          typename field_types_as_ids,
+          typename selected_field_ids,
+          typename tup_t = std::tuple<>,
+          size_t field_no = 0>
+struct generate_record_tuple_type
+{
+    //!\brief The return type.
+    using type = tup_t;
+};
+
+//!\cond
+template <typename field_types,
+          typename field_types_as_ids,
+          typename selected_field_ids,
+          typename tup_t,
+          size_t field_no>
+    requires field_no < selected_field_ids::as_array.size()
```

Comment:
Can you shortly document the rationale behind this here.

Reply:
yes